smt/expr.cpp

```diff
@@ -511,6 +512,59 @@ expr expr::lshr(const expr &rhs) const {
   return Z3_mk_bvlshr(ctx(), ast(), rhs());
 }
 
+expr expr::rotate(const expr &rhs, bool shift_left) const {
+  C(rhs);
+  if (isZero() || isAllOnes() || rhs.isZero())
+    return *this;
+
+  auto nbits = bits();
+  expr rhs_mod_width = rhs.urem(mkUInt(nbits, nbits));
+  if (rhs_mod_width.isZero())
+    return *this;
+
+  // FIXME: could do constant-folding, but *only* for i8/i16/i32/i64.
+
+  uint64_t rhs_mod_width_const;
+  if (rhs_mod_width.isUInt(rhs_mod_width_const))
+    return shift_left ? Z3_mk_rotate_left(ctx(), rhs_mod_width_const, ast())
+                      : Z3_mk_rotate_right(ctx(), rhs_mod_width_const, ast());
+
+  return shift_left ? Z3_mk_ext_rotate_left(ctx(), ast(), rhs_mod_width.ast())
+                    : Z3_mk_ext_rotate_right(ctx(), ast(), rhs_mod_width.ast());
+}
+
+expr expr::funnelShift(const expr &a, const expr &b, const expr &c,
```

Comment:
I think this is overly complicated. I would split this in fshl/fshr and just do: (a.concat(b) << (c.urem(bits).zext(bits)).extract(...).
Constant folding should happen in concat, urem, shift, etc. (even if it doesn't today for all cases; we can improve those separately)

Reply:
Hmm, okay.
And no `rotate` (`Z3_mk_{,ext}_rotate_{left,right}()`) short-circuiting either?

Reply:
No. The reason is that internally Z3 doesn't really have rotate anyway. So it's better not to use it and use plain shifts. We can add constant-folding to shifts if performance is impacted (overall; not just in these test cases).

Reply:
Ok, understood, thank you.